dice_ml/explainer_interfaces/explainer_base.py

```diff
@@ -299,20 +278,100 @@ def do_binary_search(self, diff, decimal_prec, query_instance, cf_ix, feature, f
 
         return final_cfs_sparse
 
+    def infer_target_cfs_class(self, desired_class_input, original_pred,
+            num_output_nodes):
+        """ Infer the target class for generating CFs. Only called when
+        model_type=="classifier".
+        TODO: Add support for opposite desired class in multiclass. Downstream methods should decide
+        whether it is allowed or not.
+        """
+        target_class = None
+        if desired_class_input == "opposite":
+            if num_output_nodes == 2:
+                original_pred_1 = np.argmax(original_pred)
+                target_class = int(1 - original_pred_1)
+
+            elif num_output_nodes > 2:
+                raise ValueError("Desired class cannot be opposite if the number of classes is more than 2.")
+
+        if isinstance(desired_class_input, int):
+            if desired_class_input >= 0 and desired_class_input < self.num_output_nodes:
+                target_class = desired_class_input
+            else:
+                raise ValueError("Desired class should be within 0 and num_classes-1.")
+        return target_class
+
+    def infer_target_cfs_range(self, desired_range_input):
+        target_range = None
+        if desired_range_input is None:
+            raise ValueError("Need to provide a desired_range for the target counterfactuals for a regression model.")
+        else:
+            if desired_range_input[0] > desired_range_input[1]:
+                raise ValueError("Invalid Range!")
+            else:
+                target_range = desired_range_input
+        return target_range
+
+    def decide_cf_validity(self, model_outputs):
+        validity = np.zeros(len(model_outputs), dtype=np.int32)
+        for i in range(len(model_outputs)):
+            pred = model_outputs[i]
+            if self.model.model_type == "classifier":
+                if self.num_output_nodes == 2: # binary
+                    pred_1 = pred[self.num_output_nodes-1]
+                    validity[i] = 1 if ((self.target_cf_class == 0 and pred_1<= self.stopping_threshold) or (self.target_cf_class == 1 and pred_1>= self.stopping_threshold)) else 0
+                else: # multiclass
+                    maxscore = max(pred)
+                    if math.isclose(pred[self.target_cf_class], maxscore):
+                        validity[i] = 1
+            elif self.model.model_type == "regressor":
+                if pred >= self.target_cf_range[0] and pred < self.target_cf_range[1]:
+                    validity[i] = 1
+        return validity
+
+    def is_cf_valid(self, model_score):
+        """Check if a cf belongs to the target class or target range.
+        """
+        # Converting to single prediction if the prediction is provided as a
+        # singleton array
+        correct_dim = 1 if self.model.model_type == "classifier" else 0
+        if hasattr(model_score, "shape") and len(model_score.shape) > correct_dim:
+            model_score = model_score[0]
```

Comment:
Shouldn't this be model_score[1]?

Reply:
no, this is just to correct if the output is given as [[a,b]], in that case we just want to take [a,b]